### ISSUE TITLE
Occasional segfault in testautomation: mouse_warpMouseInWindow

### ISSUE THREAD

**Issue by madebr** [libsdl-org/SDL#10395]
A segmentation fault happened when running `testautomation --trackmem` in gdb.
The error is not always reproducible.

This was the output of the mouse test suite:
```
INFO:  07/27/24 19:33:38: ===== Test Suite 12: 'Mouse' started
INFO:  07/27/24 19:33:38: ----- Test Case 12.1: 'mouse_getMouseState' started
INFO:  07/27/24 19:33:38: Test Description: 'Check call to SDL_GetMouseState'
INFO:  07/27/24 19:33:38: Test Iteration 1: execKey 16103970059141016811
[New Thread 0x7fffe4c006c0 (LWP 396396)]
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_PumpEvents()': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetMouseState(NULL, NULL)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetMouseState(&x, NULL)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of x is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetMouseState(NULL, &y)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of y is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetMouseState(&x, &y)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of x is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of y is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=13 Passed=13 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.1 sec
INFO:  07/27/24 19:33:39: >>> Test 'mouse_getMouseState': Passed
INFO:  07/27/24 19:33:39: ----- Test Case 12.2: 'mouse_getRelativeMouseState' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_GetRelativeMouseState'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 14613464025284677728
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_PumpEvents()': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetRelativeMouseState(NULL, NULL)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetRelativeMouseState(&x, NULL)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of x is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetRelativeMouseState(NULL, &y)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of y is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetRelativeMouseState(&x, &y)': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of x is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate that value of y is > -FLT_MAX, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate state returned from function, got: 0': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=13 Passed=13 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.0 sec
INFO:  07/27/24 19:33:39: >>> Test 'mouse_getRelativeMouseState': Passed
INFO:  07/27/24 19:33:39: ----- Test Case 12.3: 'mouse_createFreeCursor' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_CreateCursor and SDL_DestroyCursor'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 16183066726033667054
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_CreateCursor()': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate result from SDL_CreateCursor() is not NULL': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_DestroyCursor()': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=3 Passed=3 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.0 sec
INFO:  07/27/24 19:33:39: >>> Test 'mouse_createFreeCursor': Passed
INFO:  07/27/24 19:33:39: ----- Test Case 12.4: 'mouse_showCursor' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_ShowCursor'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 9319797585748017996
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_CursorVisible()': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_HideCursor()': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_CursorVisible()': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate new state, expected: SDL_FALSE, got: SDL_FALSE': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_ShowCursor()': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_CursorVisible()': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate new state, expected: SDL_TRUE, got: SDL_TRUE': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=7 Passed=7 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.0 sec
INFO:  07/27/24 19:33:39: >>> Test 'mouse_showCursor': Passed
INFO:  07/27/24 19:33:39: ----- Test Case 12.5: 'mouse_setCursor' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_SetCursor'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 8965056345982556044
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_CreateCursor()': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate result from SDL_CreateCursor() is not NULL': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_SetCursor(cursor)': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_SetCursor(NULL)': Passed
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_DestroyCursor()': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=5 Passed=5 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.0 sec
INFO:  07/27/24 19:33:39: >>> Test 'mouse_setCursor': Passed
INFO:  07/27/24 19:33:39: ----- Test Case 12.6: 'mouse_getCursor' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_GetCursor'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 11105647495969089445
INFO:  07/27/24 19:33:39: Assert 'Call to SDL_GetCursor()': Passed
INFO:  07/27/24 19:33:39: Assert 'Validate result from SDL_GetCursor() is not NULL': Passed
INFO:  07/27/24 19:33:39: Assert Summary: Total=2 Passed=2 Failed=0
INFO:  07/27/24 19:33:39: Total Test runtime: 0.0 sec
INFO:  07/27/24 19:33:39: ----- Test Case 12.7: 'mouse_warpMouseInWindow' started
INFO:  07/27/24 19:33:39: Test Description: 'Check call to SDL_WarpMouseInWindow'
INFO:  07/27/24 19:33:39: Test Iteration 1: execKey 3059679517458032303
[Thread 0x7fffe4c006c0 (LWP 396396) exited]
[New Thread 0x7fffe4c006c0 (LWP 396397)]
[New Thread 0x7fffdbe006c0 (LWP 396398)]
[Thread 0x7fffe4c006c0 (LWP 396397) exited]
[Thread 0x7fffdbe006c0 (LWP 396398) exited]
[New Thread 0x7fffdbe006c0 (LWP 396399)]
[New Thread 0x7fffe4c006c0 (LWP 396400)]
[Thread 0x7fffdbe006c0 (LWP 396399) exited]
[Thread 0x7fffe4c006c0 (LWP 396400) exited]
INFO:  07/27/24 19:33:39: Assert 'SDL_CreateWindow()': Passed
INFO:  07/27/24 19:33:39: Assert 'Check SDL_CreateWindow result': Passed

Thread 1 "testautomation" received signal SIGSEGV, Segmentation fault.
```

Here's the full backtrace in case it helps:
```
(gdb) bt full
#0  0x00007ffff7a95d34 in pthread_mutex_lock@@GLIBC_2.2.5 () at /lib64/libc.so.6
#1  0x00007ffff7facfc1 in wl_proxy_destroy () at /lib64/libwayland-client.so.0
#2  0x00007ffff7e9492e in wl_callback_destroy (wl_callback=0x7ffff705f5b9)
    at /home/maarten/projects/SDL/cmake-build-debug/wayland-generated-protocols/wayland-client-protocol.h:1250
#3  0x00007ffff7e961c2 in Wayland_ShowCursor (cursor=0x5e1cd0)
    at /home/maarten/projects/SDL/src/video/wayland/SDL_waylandmouse.c:615
        vd = 0x4b67b0
        d = 0x4b62b0
        input = 0x4b6510
        pointer = 0x4bfad0
        scale = 1
#4  0x00007ffff7c7fd77 in SDL_SetCursor_REAL (cursor=0x5e1cd0)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:1621
        mouse = 0x7ffff7f95000 <SDL_mouse>
#5  0x00007ffff7c7d3b1 in SDL_SetMouseFocus (window=0x1e88380)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:529
        mouse = 0x7ffff7f95000 <SDL_mouse>
#6  0x00007ffff7c7d505 in SDL_UpdateMouseFocus (window=0x1e88380, x=75, y=100, buttonstate=0, send_mouse_motion=1)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:569
        mouse = 0x7ffff7f95000 <SDL_mouse>
        inWindow = 1
#7  0x00007ffff7c7d5b8 in SDL_SendMouseMotion (timestamp=0, window=0x1e88380, mouseID=0, relative=0, x=75, y=100)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:581
        mouse = 0x7ffff7f95000 <SDL_mouse>
#8  0x00007ffff7e96597 in Wayland_WarpMouse (window=0x1e88380, x=75, y=100)
    at /home/maarten/projects/SDL/src/video/wayland/SDL_waylandmouse.c:701
        toggle_lock = 1
        vd = 0x4b67b0
        d = 0x4b62b0
        wind = 0x7b05b0
        input = 0x4b6510
#9  0x00007ffff7c7f35a in SDL_PerformWarpMouseInWindow (window=0x1e88380, x=75, y=100, ignore_relative_mode=0)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:1268
        mouse = 0x7ffff7f95000 <SDL_mouse>
#10 0x00007ffff7c7f44f in SDL_WarpMouseInWindow_REAL (window=0x1e88380, x=75, y=100)
    at /home/maarten/projects/SDL/src/events/SDL_mouse.c:1291
        mouse = 0x7ffff7f95000 <SDL_mouse>
#11 0x00007ffff7c6c7e0 in SDL_WarpMouseInWindow (a=0x1e88380, b=75, c=100)
    at /home/maarten/projects/SDL/src/dynapi/SDL_dynapi_procs.h:904
#12 0x000000000041e272 in mouse_warpMouseInWindow (arg=0x0)
    at /home/maarten/projects/SDL/test/testautomation_mouse.c:464
        w = 320
        h = 200
        numPositions = 6
        xPositions = {-1, 0, 1, 319, 320, 321}
        yPositions = {-1, 0, 1, 199, 200, 201}
        x = 75
        y = 100
        i = 3600000
        j = 184
        window = 0x1e88380
#13 0x00000000004586b7 in SDLTest_RunTest
    (testSuite=0x49dc00 <mouseTestSuite>, testCase=0x47cdc0 <mouseTest7>, execKey=3059679517458032303, forceTestRun=0) at /home/maarten/projects/SDL/src/test/SDL_test_harness.c:270
        timer = 184
        testCaseResult = 0
        testResult = 0
        fuzzerCount = 5
#14 0x0000000000459090 in SDLTest_RunSuites
    (testSuites=0x49ce20 <testSuites>, userRunSeed=0x0, userExecKey=0, filter=0x0, testIterations=1)
    at /home/maarten/projects/SDL/src/test/SDL_test_harness.c:590
        totalNumberOfTests = 347
        failedNumberOfTests = 0
        suiteCounter = 12
        testCounter = 7
        iterationCounter = 1
        testSuite = 0x49dc00 <mouseTestSuite>
        testCase = 0x47cdc0 <mouseTest7>
        runSeed = 0x7fffffffd9b0 "JMC8ALHN5OMPSUE7"
        currentSuiteName = 0x47d040 "Mouse"
        currentTestName = 0x47cd80 "mouse_warpMouseInWindow"
        execKey = 3059679517458032303
        runStartSeconds = 137449.609
        suiteStartSeconds = 137458.469
        testStartSeconds = 137458.531
        runEndSeconds = 1.40129846e-45
        suiteEndSeconds = 137458.469
        testEndSeconds = 137458.531
        runtime = 0
        suiteFilter = 0
        suiteFilterName = 0x0
        testFilter = 0
        testFilterName = 0x0
        forceTestRun = 0
        testResult = 0
        runResult = 0
        totalTestFailedCount = 0
        totalTestPassedCount = 178
        totalTestSkippedCount = 2
        testFailedCount = 0
        testPassedCount = 6
        testSkippedCount = 0
        countSum = 116
        failedTests = 0x6b1340
        generatedSeed = "JMC8ALHN5OMPSUE7"
#15 0x000000000040a259 in main (argc=2, argv=0x7fffffffdcb8)
    at /home/maarten/projects/SDL/test/testautomation.c:161
        result = 0
        testIterations = 1
        userExecKey = 0
        userRunSeed = 0x0
        filter = 0x0
        i = 0
        done = 0
        event = {type = 0, common = {type = 0, reserved = 0, timestamp = 0}, display = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, displayID = 0, data1 = 0, data2 = 0}, window = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, data1 = 0, data2 = 0}, kdevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0}, key = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, scancode = SDL_SCANCODE_UNKNOWN, key = 0, mod = 0, raw = 0, state = 0 '\000', repeat = 0 '\000'}, edit = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, text = 0x0, start = 0, length = 0}, edit_candidates = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, candidates = 0x0, num_candidates = 0, selected_candidate = 0, horizontal = 0}, text = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, text = 0x0}, mdevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0}, motion = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, state = 0, x = 0, y = 0, xrel = 0, yrel = 0}, button = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, button = 0 '\000', state = 0 '\000', clicks = 0 '\000', padding = 0 '\000', x = 0, y = 0}, wheel = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, x = 0, y = 0, direction = SDL_MOUSEWHEEL_NORMAL, mouse_x = 0, mouse_y = 0}, jdevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0}, jaxis = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, axis = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', value = 0, padding4 = 0}, jball = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, ball = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', xrel = 0, yrel = 0}, jhat = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, hat = 0 '\000', value = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, jbutton = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, button = 0 '\000', state = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, jbattery = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, state = SDL_POWERSTATE_UNKNOWN, percent = 0}, gdevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0}, gaxis = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, axis = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', value = 0, padding4 = 0}, gbutton = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, button = 0 '\000', state = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, gtouchpad = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, touchpad = 0, finger = 0, x = 0, y = 0, pressure = 0}, gsensor = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, sensor = 0, data = {0, 0, 0}, sensor_timestamp = 0}, adevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, recording = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000'}, cdevice = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0}, sensor = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, which = 0, data = {0, 0, 0, 0, 0, 0}, sensor_timestamp = 0}, quit = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0}, user = {type = 0, reserved = 0, timestamp = 0, windowID = 0, code = 0, data1 = 0x0, data2 = 0x0}, tfinger = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, touchID = 0, fingerID = 0, x = 0, y = 0, dx = 0, dy = 0, pressure = 0, windowID = 0}, ptip = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, tip = 0 '\000', state = 0 '\000', pen_state = 0, x = 0, y = 0, axes = {0, 0, 0, 0, 0, 0}}, pmotion = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, padding1 = 0 '\000', padding2 = 0 '\000', pen_state = 0, x = 0, y = 0, axes = {0, 0, 0, 0, 0, 0}}, pbutton = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, which = 0, button = 0 '\000', state = 0 '\000', pen_state = 0, x = 0, y = 0, axes = {0, 0, 0, 0, 0, 0}}, drop = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0, windowID = 0, x = 0, y = 0, source = 0x0, data = 0x0}, clipboard = {type = SDL_EVENT_FIRST, reserved = 0, timestamp = 0}, padding = '\000' <repeats 127 times>}
        list = 0
```

Using 1d5f9c7b92225afac8731c12a159af133a8a5071

System: fedora 39 x86_64

**Comment by madebr** [comment]
Running `testautomation --filter mouse --iterations 10000`, I can reproduce this, and similar, issues.
I am also seeing a hang (deadlock?) in the same test.

**Comment by slouken** [comment]
@Kontrabant, is this related to your mouse warp changes?

**Comment by Kontrabant** [comment]
Hmm, the exact place where it’s crashing wasn’t touched in the warp work. The backtrace looks like it’s dereferencing a garbage pointer.  I’ll have to investigate further.

**Comment by Kontrabant** [comment]
Ah, I think I see what is happening. A removed cursor visibility variable in the Wayland code was also stealthily acting as a “was initialized” flag. This check will need to be restored.

**Comment by Kontrabant** [comment]
Now that I'm back home and trying this, I can reproduce this even if I roll back to before the mouse warp changes. For some reason, the internal cursor data is being set to garbage when running the mouse tests sequentially and the warp test is hit. Hmm…

**Comment by Kontrabant** [comment]
Ah, found the culprit. If no window had the mouse focus, the Wayland function to retrieve the system cursor could fail, leaving the cursor in a weird state. I'm genuinely surprised this has never come up before, because that code has been that way for years.

**Comment by Kontrabant** [comment]
Fixed and backported to SDL2 and 2.30.x
